### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Comma): subcategory defined by morphism properties

### PR DESCRIPTION
**Pull request by chrisflav**
Given functors `L : A ⥤ T` and `R : B ⥤ T` and morphism properties `P`, `Q` and `W`
on `T`, `A` and `B` respectively, we define the subcategory `P.Comma L R Q W` of
`Comma L R` where

- objects are objects of `Comma L R` with the structural morphism satisfying `P`, and
- morphisms are morphisms of `Comma L R` where the left morphism satisfies `Q` and the
  right morphism satisfies `W`.

For an object `X : T`, this specializes to `P.Over Q X` which is the subcategory of `Over X`
where the structural morphism satisfies `P` and where the horizontal morphisms satisfy `Q`.
Common examples of the latter are e.g. the category of schemes étale (finite, affine, etc.)
over a base `X`. Here `Q = ⊤`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/MorphismProperty/Comma.lean [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+@[simps]
+def Hom.comp {X Y Z : P.Comma L R Q W} (f : Comma.Hom X Y) (g : Comma.Hom Y Z) :
+    Comma.Hom X Z where
+  left := f.left ≫ g.left
+  right := f.right ≫ g.right
+  prop_hom_left := Q.comp_mem _ _ f.prop_hom_left g.prop_hom_left
+  prop_hom_right := W.comp_mem _ _ f.prop_hom_right g.prop_hom_right
+
+variable (L R P Q W) in
+instance : Category (P.Comma L R Q W) where
```
Comment (joelriou) [inline Mathlib/CategoryTheory/MorphismProperty/Comma.lean:101]: `ContainsIdentities` or `IsStableUnderComposition` are the sharper assumptions for the definitions `id` and `comp` above, but I would suggest using `IsMultiplicative` as a single assumption for the definition of the category instance.

File: Mathlib/CategoryTheory/MorphismProperty/Comma.lean [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+end Comma
+
+end Comma
+
+section Over
+
+variable {T : Type*} [Category T]
+
+/-- Given a morphism property `P` on a category `C` and an object `X : C`, this is the
+subcategory of `Over X` defined by `P` where morphisms satisfy `Q`. -/
```
Comment (erdOne) [inline Mathlib/CategoryTheory/MorphismProperty/Comma.lean:200]: Is this the right way around?
Or is this "subcategory of `Over X` defined by `Q` where morphisms satisfy `P`"

File: Mathlib/CategoryTheory/MorphismProperty/Comma.lean [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+protected abbrev Over (P Q : MorphismProperty T) (X : T) : Type _ :=
+  P.Comma (Functor.id T) (Functor.fromPUnit X) Q ⊤
+
+variable (P Q : MorphismProperty T) [Q.IsMultiplicative] (X : T)
+
+/-- Construct a morphism in `P.Over Q X` from a morphism in `Over.X`. -/
```
Comment (erdOne) [inline Mathlib/CategoryTheory/MorphismProperty/Comma.lean:206]:
```suggestion
/-- Construct a morphism in `P.Over Q X` from a morphism in `Over X`. -/
```

File: Mathlib/CategoryTheory/MorphismProperty/Comma.lean [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+  Comma.forget (Functor.id T) (Functor.fromPUnit X) P Q ⊤
+
+end Over
+
+section Under
+
+variable {T : Type*} [Category T]
+
+/-- Given a morphism property `P` on a category `C` and an object `X : C`, this is the
+subcategory of `Under X` defined by `P` where morphisms satisfy `Q`. -/
```
Comment (erdOne) [inline Mathlib/CategoryTheory/MorphismProperty/Comma.lean:224]: Same here.

File: Mathlib/CategoryTheory/MorphismProperty/Comma.lean (hunk 2 above)
Comment (chrisflav) [reply Mathlib/CategoryTheory/MorphismProperty/Comma.lean:200]: This is the subcategory of `Over X` where the structural morphisms satisfy `P` and for `X Y : P.Over Q X` a morphism from `X` to `Y` is a morphism `f` in `Over X` where `f.left` satisfies `Q`. Does this clarify it?

Comment (erdOne) [reply Mathlib/CategoryTheory/MorphismProperty/Comma.lean:200]: Oh then the original wording is right. I was just confused.

**Comment by erdOne** [comment]
I was also thinking about how to do `k`-Schemes and I don't think `Over` categories are the best way of doing it. 
But we will need this anyways to define etale sites etc.

Thanks!
maintainer merge

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**